### PR TITLE
added switches to skip phases

### PR DESCRIPTION
**Pull request by dogeared**

### CODE REVIEW

File: bin/migrate.js [hunk 1]
```diff
@@ -32,12 +32,37 @@ async function migrate() {
     const timer = logger.startTimer();
 
     await introspect();
-    await migrateCustomSchema();
```
Comment (rchild-okta) [inline bin/migrate.js]: Like the idea in general, we should add this to the README and provide some guidance on the implications of skipping steps (noted below).

File: bin/migrate.js [hunk 2]
```diff
@@ -32,12 +32,37 @@ async function migrate() {
...
-    await migrateAccounts();
-    await migrateDirectories();
-    await migrateGroups();
-    await migrateOrganizations();
-    await migrateApplications();
+
+    if (!config.skipCustomSchema) {
```
Comment (rchild-okta) [inline bin/migrate.js]: This is okay once they've processed all accounts and we've pushed a schema with all known properties. If properties change or don't exist on the Okta user, that user create/update will fail.

File: bin/migrate.js [hunk 3]
```diff
@@ -32,12 +32,37 @@ async function migrate() {
...
+      await migrateCustomSchema();
+    } else {
+      logger.header('skipping Custom Schema Phase');
+    }
+    if (!config.skipAccounts) {
```
Comment (rchild-okta) [inline bin/migrate.js]: This should be okay since we store the Okta userId in the tmp/ files. If they add users since the last run, skipping this won't add them, and you'll see errors downstream if there are any group mappings, etc.

File: bin/migrate.js [hunk 4]
```diff
@@ -32,12 +32,37 @@ async function migrate() {
...
+      await migrateAccounts();
+    } else {
+        logger.header('skipping Accounts Phase');
+    }
+    if (!config.skipDirectories) {
```
Comment (rchild-okta) [inline bin/migrate.js]: This will be okay if there are no buckets that use the created directory. For example, if a directory created here is an accountProvider for an organization, it will fail downstream because we don't have a reference to the Okta group that's created from this directory.

File: bin/migrate.js [hunk 5]
```diff
@@ -32,12 +32,37 @@ async function migrate() {
...
+      await migrateDirectories();
+    } else {
+        logger.header('skipping Directories Phase');
+    }
+    if (!config.skipGroups) {
```
Comment (rchild-okta) [inline bin/migrate.js]: Same here - it's fine if the downstream operations don't depend on the id of the group that's created.

File: util/config.js [hunk 6]
```diff
@@ -80,6 +80,36 @@ const config = yargs
...
+        required: false,
+        default: false
+    },
+    skipGroups: {
+        description: 'Skip the Groups phase',
+        required: false,
+        default: false
+    },
+    skipOrganizations: {
+        description: 'skip the Organizations phase',
```
Comment (rchild-okta) [inline util/config.js]: Nit: Capitalize skip in this description and the next (let's be consistent). As an aside, it might be clearer to say "Skip importing Stormpath Organizations" rather than use the word "phase".

Comment (dogeared) [inline util/config.js]: makes sense. I'll update the wording for all of them. And, I will add verbiage to the README with the notes you gave for each skip AND stating that you skip at your own risk. ;)

File: README.md [hunk 7]
```diff
@@ -159,6 +159,39 @@ Logging level. Defaults to `info`.
 - Options: `error`, `warn`, `info`, `verbose`, `debug`, `silly`
 - Example: `--logLevel verbose`
 
+### Skipping
+
+If the tool crashes along the way or you know that only certain types of objects have been updated in the latest export, you can skip certain stages of the import.
+
+*Note*: These are advanced options. Generally speaking, each stage of the import depends on the previous stage having completed. If you skip an import stage and dependent
```
Comment (rchild-okta) [inline README.md]: nit: no need for a newline here

File: README.md [hunk 8]
```diff
@@ -159,6 +159,39 @@ Logging level. Defaults to `info`.
...
+data is not present, you will see errors during the import.
+
+Here's are the skip switches (they are false by default):
```
Comment (rchild-okta) [inline README.md]: Here are

File: README.md [hunk 9]
```diff
@@ -159,6 +159,39 @@ Logging level. Defaults to `info`.
...
+
+### `--skipCustomSchema`
+
+Skip custom schema create & update. This is okay once all accounts have previously been processed and a schema has been pushed with all know properties.
```
Comment (rchild-okta) [inline README.md]: nit: not sure we need the word "previously"

know -> known

File: README.md [hunk 10]
```diff
@@ -159,6 +159,39 @@ Logging level. Defaults to `info`.
...
+
+### `--skipAccounts`
+
+Skip account create & update. Any new or updated accounts in the export are ignored.
+
+### `--skipDirectories`  
+
+Skip directory create & update. This is ok as long as there aren't any new account providers that would cause downstream errors because we don't have a reference to them.
```
Comment (rchild-okta) [inline README.md]: If your directories are account providers for other things downstream, those things won't work. I.e. if the directory is an accountProvider for an organization, we will not create any of the memberships between that directory and the organization group that is created.

tldr; it's not just new directories, it's all of them.

File: README.md [hunk 11]
```diff
@@ -159,6 +159,39 @@ Logging level. Defaults to `info`.
...
+
+### `--skipGroups`
+
+Skip group create and update. This is ok as long as there aren't any new groups that would cause downstream errors because we don't have a reference to them.
```
Comment (rchild-okta) [inline README.md]: Same here. You can think of it this way:

1. In the group phase, for each stormpath group we look up the group in Okta. If it exists, we get its id and put it in a map of stormpath groupId -> okta groupId. If it doesn't exist, we create it and then add that new id to the same map.

2. Later on, when we have another bucket that uses the group as an accountProvider, we use that map to assign group memberships.

File: README.md [hunk 12]
```diff
@@ -159,6 +159,39 @@ Logging level. Defaults to `info`.
...
+
+### `--skipOrganizations`
+
+Skip organization create and update. This is ok as long as there aren't any new account providers that would cause downstream errors because we don't have a reference to them.
```
Comment (rchild-okta) [inline README.md]: Reading through these, it might be better to just be explicit about what will break:

1. Organizations have an orgAccountStoreMap that can include groups or directories. If you skip groups or directories that are used by an organization, it won't work (won't assign all group memberships that are expected).

2. Applications have an accountStoreMap that can include groups, organizations, or directories. If you skip groups, organizations, or directories that are used by an application, it won't work.
